bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/WorkbookEditorsHandler.java

```diff
@@ -161,46 +162,56 @@ private List<EditorReference> getParts(WorkbenchPage page) {
 	 */
 	private Map<EditorReference, String> generateColumnLabelTexts(List<EditorReference> editorReferences) {
 		Map<EditorReference, String> editorReferenceLabelTexts = new HashMap<>(editorReferences.size());
-		Map<String, List<Entry<EditorReference, IPath>>> collisionsMap = new HashMap<>(editorReferences.size());
-		editorReferences.forEach(editorReference -> {
-			try {
-				IPathEditorInput iPathEditorInput = Adapters.adapt(editorReference.getEditorInput(),
-						IPathEditorInput.class);
-				if (iPathEditorInput != null && iPathEditorInput.getPath() != null) {
-					IPath path = iPathEditorInput.getPath();
-
-					List<Entry<EditorReference, IPath>> referencesWithSameTitle = collisionsMap
-							.get(editorReference.getTitle());
-					if (referencesWithSameTitle == null) {
-						referencesWithSameTitle = new ArrayList<>();
-						collisionsMap.put(editorReference.getTitle(), referencesWithSameTitle);
+		Map<String, List<EditorReference>> collisionsMap = editorReferences.stream()
+				.collect(Collectors.groupingBy(r -> r.getName()));
+
+		for (Entry<String, List<EditorReference>> groupedEditorReferences : collisionsMap.entrySet()) {
+			if (groupedEditorReferences.getValue().size() == 1) {
+				groupedEditorReferences.getValue().stream().forEach(editorReference -> editorReferenceLabelTexts
+						.put(editorReference, getWorkbenchPartReferenceText(editorReference)));
+			} else {
+				List<Entry<EditorReference, IPath>> refsToMakeDistinguishableViaPathSegments = new ArrayList<>();
+				for (EditorReference editorReference : groupedEditorReferences.getValue()) {
+					try {
+						// we only detect collisions for IPathEditorInput and only if the name used by
+						// the editor reference is the filename. Otherwise, this would break scenarios
+						// where editors override the name used, e.g. for virtual file systems using
+						// org.eclipse.core.internal.filesystem.FileCache
+						IPathEditorInput iPathEditorInput = Adapters.adapt(editorReference.getEditorInput(),
+								IPathEditorInput.class);
+						IPath path;
+						if (iPathEditorInput != null && (path = iPathEditorInput.getPath()) != null && path.toFile()
+								.toPath().getFileName().toString().equals(groupedEditorReferences.getKey())) {
```

Comment:
I think you can use `path.lastSegment()` here instead of converting to a nio `Path`.

Reply:
Good point. I was confused why code completion didn't offer this, but I only checked for `.get...()` methods :)